be/src/vec/columns/column_dictionary.h

```diff
@@ -0,0 +1,381 @@
+    }
+
+    void get_extremes(Field& min, Field& max) const {
+        LOG(FATAL) << "get_extremes not supported in ColumnDictionary";
+    }
+
+    bool can_be_inside_nullable() const override { return true; }
+
+    bool is_fixed_and_contiguous() const override { return true; }
+
+    size_t size_of_value_if_fixed() const override { return sizeof(T); }
+
+    [[noreturn]] StringRef get_raw_data() const override {
+        LOG(FATAL) << "get_raw_data not supported in ColumnDictionary";
+    }
+
+    [[noreturn]] bool structure_equals(const IColumn& rhs) const override {
+        LOG(FATAL) << "structure_equals not supported in ColumnDictionary";
+    }
+
+    [[noreturn]] ColumnPtr filter(const IColumn::Filter& filt,
+                                  ssize_t result_size_hint) const override {
+        LOG(FATAL) << "filter not supported in ColumnDictionary";
+    };
+
+    [[noreturn]] ColumnPtr permute(const IColumn::Permutation& perm, size_t limit) const override {
+        LOG(FATAL) << "permute not supported in ColumnDictionary";
+    };
+
+    [[noreturn]] ColumnPtr replicate(const IColumn::Offsets& replicate_offsets) const override {
+        LOG(FATAL) << "replicate not supported in ColumnDictionary";
+    };
+
+    [[noreturn]] MutableColumns scatter(IColumn::ColumnIndex num_columns,
+                                        const IColumn::Selector& selector) const override {
+        LOG(FATAL) << "scatter not supported in ColumnDictionary";
+    }
+
+    Status filter_by_selector(const uint16_t* sel, size_t sel_size, IColumn* col_ptr) override {
+        auto* res_col = reinterpret_cast<vectorized::ColumnString*>(col_ptr);
+        for (size_t i = 0; i < sel_size; i++) {
+            uint16_t n = sel[i];
+            auto& code = reinterpret_cast<T&>(codes[n]);
+            auto value = dict.get_value(code);
+            res_col->insert_data(value.ptr, value.len);
+        }
+        return Status::OK();
+    }
+
+    void replace_column_data(const IColumn&, size_t row, size_t self_row = 0) override {
+        LOG(FATAL) << "should not call replace_column_data in ColumnDictionary";
+    }
+
+    void replace_column_data_default(size_t self_row = 0) override {
+        LOG(FATAL) << "should not call replace_column_data_default in ColumnDictionary";
+    }
+
+    void insert_many_dict_data(const int32_t* data_array, size_t start_index,
+                               const StringRef* dict_array, size_t data_num,
+                               uint32_t dict_num) override {
+        if (!is_dict_inited()) {
+            dict.reserve(dict_num);
+            for (uint32_t i = 0; i < dict_num; ++i) {
+                auto value = StringValue(dict_array[i].data, dict_array[i].size);
+                dict.insert_value(value);
+            }
+            dict_inited = true;
+        }
+
+        for (int i = 0; i < data_num; i++, start_index++) {
+            int32_t code = data_array[start_index];
+            insert_data(code);
```

Comment:
Using memcpy here can improve performance;
Now writing to this ColumnDictionary must be very slow

Reply:
Great idea, I'll optimize this right away